### PR TITLE
Puppet 4x support

### PR DESCRIPTION
**Pull request by Fabian1976**
These main things are changed:
 - Function type is no longer supported. Modified it to type3x.
 - File mode needed to be a string. Not an int
 - Variables in templates needed to have an @ in front of them
 - $FW variable needed to be lower case
 - inline_template removed on several locations

I have tested the installation of ipa-server, the creation of a user and adding of a service.
Not tested:
 - Client installation
 - Replica
 - Adding a host

There is still an issue when running the puppet agent a second time. The puppet code "thinks" it needs to change the user on every run but that fails with "/returns: ipa: ERROR: no modifications to be performed". Probably an error in diff.py.
When running diff.py manually, i get the following error:
```
./diff.py user ntesla --first='Nikola' --last='Tesla' --cn='Nikola Tesla' --displayname='Nikola Tesla' --initials='NT' --email='ntesla@core.cmc.lan' --gecos='Nikola Tesla' --shell='/bin/sh' --homedir='/home/ntesla'
ipa: INFO: trying https://puppetdev-fabian02.core.cmc.lan/ipa/session/xml
ipa: INFO: Forwarding 'user_show' to json server 'https://puppetdev-fabian02.core.cmc.lan/ipa/session/xml'
Traceback (most recent call last):
  File "./diff.py", line 531, in <module>
    output = ipalib.api.Command.user_show(unicode(args.primarykey), all=True)
  File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 442, in __call__
    ret = self.run(*args, **options)
  File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 760, in run
    return self.forward(*args, **options)
  File "/usr/lib/python2.7/site-packages/ipalib/plugins/user.py", line 825, in forward
    return super(user_show, self).forward(*keys, **options)
  File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 781, in forward
    return self.Backend.rpcclient.forward(self.name, *args, **kw)
  File "/usr/lib/python2.7/site-packages/ipalib/rpc.py", line 885, in forward
    command = getattr(self.conn, name)
  File "/usr/lib/python2.7/site-packages/ipalib/backend.py", line 97, in __get_conn
    self.id, threading.currentThread().getName())
```

But the basic installation of ipa-server works fine.

### CODE REVIEW

File: manifests/server.pp [hunk 1]
```diff
@@ -71,7 +71,7 @@
 	$peer_excludes = [],		# never purge these peer excludes...
 	$ensure = present		# TODO: support uninstall with 'absent'
 ) {
-	$FW = '$FW'			# make using $FW in shorewall easier...
+	$fw = '$FW'			# make using $fw in shorewall easier...
```
Comment (purpleidea) [inline manifests/server.pp]: Nitpick, technically the comment was correct.

File: manifests/server.pp [hunk 2]
```diff
@@ -644,46 +644,46 @@
 		#ACTION      SOURCE DEST                PROTO DEST  SOURCE  ORIGINAL
 		#                                             PORT  PORT(S) DEST
 		shorewall::rule { 'http': rule => "
-		HTTP/ACCEPT  ${net}    $FW
+		HTTP/ACCEPT  ${net}    $fw
```
Comment (purpleidea) [inline manifests/server.pp]: nitpick, technically we could use ${fw} everywhere now to make it clearer. Optional.

File: manifests/server/user.pp [hunk 3]
```diff
@@ -191,36 +191,36 @@
...
 		'boolean' => $email ? {
 			false => '',	# unmanaged
-			default => ["${valid_login}@${default_email_domain}"],	# sensible default
+			default => "${valid_login}@${default_email_domain}",	# sensible default
 		},
 		default => '',	# unmanaged
 	}
-	$args06 = type($valid_email) ? {
-		'array' => inline_template('<% if valid_email == [] %>--email=<% else %><%= valid_email.map {|x| "--email=\'"+x+"\'" }.join(" ") %><% end %>'),
+	$args06 = type3x($valid_email) ? {
```
Comment (purpleidea) [inline manifests/server/user.pp]: if $valid_email is an array, this doesn't work anymore...

File: manifests/server.pp (hunk 2 above)
Comment (Fabian1976) [inline manifests/server.pp]: fixed as requested

File: manifests/server/user.pp (hunk 3 above)
Comment (Fabian1976) [inline manifests/server/user.pp]: You're right, but that goes for a lot of parameters. According to the docs (ipa user-add --help) email should be string.
  --email=STR           Email address

**Comment by purpleidea** [comment]
Hey @Fabian1976 I've decided to merge this!

Thanks for your patch, if you'd like to keep working on this, I can add you as a committer. LMK

Cheers

**Comment by Fabian1976** [comment]
Hi James,

 

I am planning on expanding the code to be able to manage sudo rules and host/user groups and stuff like that.

So that would ben ice.

 

Kind regards,

 

Fabian.

 

Van: James <notifications@github.com>
Beantwoorden - Aan: purpleidea/puppet-ipa <reply@reply.github.com>
Datum: woensdag 26 oktober 2016 01:16
Aan: purpleidea/puppet-ipa <puppet-ipa@noreply.github.com>
CC: Fabian van der Hoeven <fabian.vanderhoeven@vermont24-7.com>, Mention <mention@noreply.github.com>
Onderwerp: Re: [purpleidea/puppet-ipa] Puppet 4x support (#9)

 

Hey @Fabian1976 I've decided to merge this!

Thanks for your patch, if you'd like to keep working on this, I can add you as a committer. LMK

Cheers

**Comment by purpleidea** [comment]
@Fabian1976 Okay, congrats, I've added you as a collaborator.

1) No merge commits please. Rebase patches...
2) At least for the first few patches let's do a two ACK system.

Does that work for you?

**Pull request merged**